eth/bor_checkpoint_verifier.go

```diff
@@ -2,59 +2,120 @@ package eth
+
+	// errRootHash is returned when we aren't able to calculate the root hash
+	// locally for a range of blocks.
+	errRootHash = errors.New("failed to get local root hash")
+
+	// errRootHashMismatch is returned when the local root hash doesn't match
+	// with the root hash in checkpoint/milestone.
+	errRootHashMismatch = errors.New("roothash mismatch")
+
+	// errEndBlock is returned when we're unable to fetch a block locally.
+	errEndBlock = errors.New("failed to get end block")
+)
+
+type borVerifier struct {
+	verify func(ctx context.Context, eth *Ethereum, handler *ethHandler, start uint64, end uint64, rootHash string) (string, error)
 }
 
-func newCheckpointVerifier(verifyFn func(ctx context.Context, handler *ethHandler, checkpoint *checkpoint.Checkpoint) (string, error)) *checkpointVerifier {
+func newBorVerifier(verifyFn func(ctx context.Context, eth *Ethereum, handler *ethHandler, start uint64, end uint64, rootHash string) (string, error)) *borVerifier {
 	if verifyFn != nil {
-		return &checkpointVerifier{verifyFn}
+		return &borVerifier{verifyFn}
 	}
 
-	verifyFn = func(ctx context.Context, handler *ethHandler, checkpoint *checkpoint.Checkpoint) (string, error) {
-		var (
-			startBlock = checkpoint.StartBlock.Uint64()
-			endBlock   = checkpoint.EndBlock.Uint64()
-		)
+	verifyFn = func(ctx context.Context, eth *Ethereum, handler *ethHandler, start uint64, end uint64, rootHash string) (string, error) {
+		var hash string
 
-		// check if we have the checkpoint blocks
+		// check if we have the given blocks
 		head := handler.ethAPI.BlockNumber()
-		if head < hexutil.Uint64(endBlock) {
-			log.Debug("Head block behind checkpoint block", "head", head, "checkpoint end block", endBlock)
-			return "", errMissingCheckpoint
+		if head < hexutil.Uint64(end) {
+			log.Debug("Head block behind given block", "head", head, "end block", end)
+			return hash, errMissingBlocks
 		}
 
-		// verify the root hash of checkpoint
-		roothash, err := handler.ethAPI.GetRootHash(ctx, startBlock, endBlock)
+		// verify the root hash
+		localRoothash, err := handler.ethAPI.GetRootHash(ctx, start, end)
 		if err != nil {
-			log.Debug("Failed to get root hash of checkpoint while whitelisting", "err", err)
-			return "", errRootHash
+			log.Debug("Failed to get root hash of given block range while whitelisting", "start", start, "end", end, "err", err)
+			return hash, errRootHash
 		}
 
-		if roothash != checkpoint.RootHash.String()[2:] {
-			log.Warn("Checkpoint root hash mismatch while whitelisting", "expected", checkpoint.RootHash.String()[2:], "got", roothash)
-			return "", errCheckpointRootHashMismatch
+		//nolint
+		if localRoothash != rootHash {
+
+			log.Warn("Root hash mismatch while whitelisting", "expected", localRoothash, "got", rootHash)
+
+			ethHandler := (*ethHandler)(eth.handler)
+
+			var (
+				rewindTo uint64
+				doExist  bool
+			)
+
+			if doExist, rewindTo, _ = ethHandler.downloader.GetWhitelistedMilestone(); doExist {
```

Comment:
I believe this part should be extracted as a function with a batch of units on it.

Reply:
Do you mean to put this if else if else in one function? And use that function here?